astropy/cosmology/core.py

```diff
@@ -1319,18 +1322,14 @@ def angular_diameter_distance_z1z2(self, z1, z2):
         if Ok0 < 0:
             raise CosmologyError('Ok0 must be >= 0 to use this method.')
 
-        outscalar = False
-        if not isiterable(z1) and not isiterable(z2):
-            outscalar = True
-
-        z1 = np.atleast_1d(z1)
-        z2 = np.atleast_1d(z2)
+        z1 = np.array(z1)
+        z2 = np.array(z2)
+        # Not elegant but try: np.broadcast(z1, z2) doubles the runtime.
+        if z1.size != z2.size and z1.size != 1 and z2.size !=1:
+            raise ValueError('z1 and z2 must have compatible sizes.')
 
-        if z1.size != z2.size:
-            raise ValueError('z1 and z2 must be the same size.')
-
-        if (z1 > z2).any():
-            raise ValueError('z2 must greater than z1')
+        if np.any(z1 > z2):
```

Comment:
p.s. I think that with the above change, it is also not necessary any more to convert `z1` and `z2` into arrays -- it will be done implicitly if needed (but I didn't check this, so see if tests still pass...).